[USMON-805] Fix HTTP/2 frame remainder bug

amitslavin opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

This PR fix couple of issues we had in our current support for HTTP/2:
- A use case in the get_first_frame process was identified where it was not possible and may have created issues, especially in edge cases with large payloads.
- There was a small bug related to frame_state->remainder, which was not handled correctly.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

Improves our accuracy for HTTP/2 in both plain and TLS support.

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

amitslavin committed:
fix large data payload issue

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -438,7 +438,7 @@ static __always_inline void tls_fix_header_frame(tls_dispatcher_arguments_t *inf
 static __always_inline bool tls_get_first_frame(tls_dispatcher_arguments_t *info, frame_header_remainder_t *frame_state, http2_frame_t *current_frame, http2_telemetry_t *http2_tel) {
     // No state, try reading a frame.
-    if (frame_state == NULL) {
+    if (frame_state == NULL || (frame_state->remainder == 0 && frame_state->header_length == 0)) {
```
modify the comment
open a jira ticket to investigate why do we have a frame state where both remainder and header_length are zero

amitslavin replied:
https://datadoghq.atlassian.net/browse/USMON-797

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -625,6 +619,10 @@ int uprobe__http2_tls_handle_first_frame(struct pt_regs *ctx) {
     // skip HTTP2 magic, if present
     tls_skip_preface(&dispatcher_args_copy);
+    if (dispatcher_args_copy.data_off == dispatcher_args_copy.data_end) {
+    // Abort early if we reached to the end of the frame (a.k.a having only the HTTP2 magic in the packet).
```
```suggestion
        // Abort early if we reached to the end of the frame (a.k.a having only the HTTP2 magic in the packet).
```

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -488,28 +488,22 @@ static __always_inline bool tls_get_first_frame(tls_dispatcher_arguments_t *info
-
     // We failed to read a frame, if we have a remainder trying to consume it and read the following frame.
     if (frame_state->remainder > 0) {
+         if (info->data_off + frame_state->remainder > info->data_end) {
+            frame_state->remainder -= info->data_end - info->data_off;
+            info->data_off = info->data_end;
+            return false;
+        }
```
document

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding.h:
```diff
@@ -639,6 +633,10 @@ int socket__http2_handle_first_frame(struct __sk_buff *skb) {
     // skip HTTP2 magic, if present
     skip_preface(skb, &dispatcher_args_copy.skb_info);
+    if (dispatcher_args_copy.skb_info.data_off == dispatcher_args_copy.skb_info.data_end) {
+    // Abort early if we reached to the end of the frame (a.k.a having only the HTTP2 magic in the packet).
```
```suggestion
        // Abort early if we reached to the end of the frame (a.k.a having only the HTTP2 magic in the packet).
```

amitslavin committed:
fix cr notes

amitslavin committed:
fix cr note

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/fix-larage-payload-size-bug

amitslavin committed:
fix cleanup remainder bug

amitslavin committed:
added uts to validate the remainder map size

guyarb commented on pkg/network/usm/usm_http2_monitor_test.go:
```diff
@@ -1379,6 +1379,80 @@ func (s *usmHTTP2Suite) TestDynamicTable() {
 	}
 }
 
+func (s *usmHTTP2Suite) TestRemainderTable() {
```
Let's document the purpose of the test

guyarb commented on pkg/network/usm/usm_http2_monitor_test.go:
```diff
@@ -1379,6 +1379,80 @@ func (s *usmHTTP2Suite) TestDynamicTable() {
+				b := newFramer().writeData(t, 1, true, []byte("test12345")).bytes()
+				message := append(a, b[11:]...)
+				return [][]byte{
+					// we split it in 10 bytes in order to split the payload itself.
```
rephrase the comment

guyarb commented on pkg/network/usm/usm_http2_monitor_test.go:
```diff
@@ -1379,6 +1379,80 @@ func (s *usmHTTP2Suite) TestDynamicTable() {
+			mapSize: 0,
+		},
+		{
+			name: "validate remainder in map",
+			// The purpose of this test is to validate that we cannot handle reassembled tcp segments.
+			messageBuilder: func() [][]byte {
+				a := newFramer().
+					writeHeaders(t, 1, usmhttp2.HeadersFrameOptions{Headers: testHeaders()}).
+					writeData(t, 1, true, emptyBody).bytes()
+				return [][]byte{
+					// we split it in 10 bytes in order to split the payload itself.
+					a[:10],
+					a[10:],
+				}
+			},
+			mapSize: 1,
```
out of scope, but isn't that a bug? shouldn't we clean the remainder?

amitslavin replied:
https://datadoghq.atlassian.net/browse/USMON-804

amitslavin committed:
fix cr notes

amitslavin committed:
return payload size limit

amitslavin commented:
@guyarb, I have reverted the changes to the payload size as it was causing different issues. I will address it in a separate PR. I've also updated the description and title of the PR. Could you please review it again to validate the current changes?

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -437,8 +437,8 @@ static __always_inline void tls_fix_header_frame(tls_dispatcher_arguments_t *inf
-    // No state, try reading a frame.
-    if (frame_state == NULL) {
+    // Attempting to read the initial frame in the packet, or handling a state where there is no remainder and finishing reading the current frame.
+    if (frame_state == NULL || (frame_state->remainder == 0 && frame_state->header_length == 0)) {
```
do we still need that change?
didn't we fix the case of having empty remainder in this PR?

guyarb replied:
still relevant

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding.h:
```diff
@@ -435,8 +435,8 @@ static __always_inline void fix_header_frame(struct __sk_buff *skb, skb_info_t *
-    // No state, try reading a frame.
-    if (frame_state == NULL) {
+    // Attempting to read the initial frame in the packet, or handling a state where there is no remainder and finishing reading the current frame.
+    if (frame_state == NULL || (frame_state->remainder == 0 && frame_state->header_length == 0)) {
```
same as tls version

guyarb reviewed (commented):
reviewed

amitslavin committed:
fix cr note

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -467,7 +467,7 @@ static __always_inline bool tls_get_first_frame(tls_dispatcher_arguments_t *info
 
     // Frame-header-remainder.
-    if (frame_state != NULL && frame_state->header_length == HTTP2_FRAME_HEADER_SIZE) {
+    if (frame_state == NULL) {
```
revert

amitslavin committed:
fix issue
